src/python/pants/backend/shell/target_types.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import re
+from enum import Enum
+from typing import Optional
+
+from pants.engine.addresses import Address
+from pants.engine.process import BinaryPathTest
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    Dependencies,
+    IntField,
+    InvalidFieldException,
+    Sources,
+    StringField,
+    Target,
+)
+from pants.util.enums import match
+
+
+class ShellSources(Sources):
+    # Normally, we would add `expected_file_extensions = ('.sh',)`, but Bash scripts don't need a
+    # file extension, so we don't use this.
+    uses_source_roots = False
+
+
+# -----------------------------------------------------------------------------------------------
+# `shunit2_tests` target
+# -----------------------------------------------------------------------------------------------
+
+
+class Shunit2Shell(Enum):
+    sh = "sh"
+    bash = "bash"
+    dash = "dash"
+    ksh = "ksh"
+    pdksh = "pdksh"
+    zsh = "zsh"
+
+    @classmethod
+    def parse_shebang(cls, shebang: bytes) -> Shunit2Shell | None:
+        if not shebang:
+            return None
+        first_line = shebang.splitlines()[0]
+        matches = re.match(rb"^#! *[/\w]*/(?P<program>\w+) *(?P<arg>\w*)", first_line)
+        if not matches:
+            return None
+        program = matches.group("program")
+        if program == b"env":
+            program = matches.group("arg")
+        try:
+            return cls(program.decode())
+        except ValueError:
+            return None
+
+    @property
+    def binary_path_test(self) -> BinaryPathTest | None:
+        arg = match(
+            self,  # type: ignore[misc]
+            {
+                self.sh: None,
```

Comment:
These exist on docker images / can be installed on them to figure out what to use for a test - perhaps a TODO to track a follow-up.

Reply:
I looked at man pages to figure these out, along with some google searches. I didn't do `sh` because this makes it sound too non-deterministic:

```
sh is a POSIX-compliant command interpreter (shell).  It is implemented by re-execing as either bash(1),
     dash(1), or zsh(1) as determined by the symbolic link located at /private/var/select/sh.  If
     /private/var/select/sh does not exist or does not point to a valid shell, sh will use one of the supported
     shells.
```

The others did not have `-v` options and required things like using a package manager to get the version.

Reply:
On the ubuntu:20.04 image stock ksh at any rate I get `--version` working:
```
root@a7fdd80769e2:/# ksh --version
  version         sh (AT&T Research) 2020.0.0
```

Reply:
Oh that works on macOS stock too. Thanks